<img src="https://images.unsplash.com/photo-1613771404721-1f92d799e49f?ixlib=rb-4.1.0&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&q=80&w=1738" width="400" height="300">

Photo by [Thimo Pedersen](www.google.com) on [Unsplash](www.google.com).

# Train Test Split Practice
## Pokemon Dataset
We are going to predictive modelling with OLS models to see what categories about a Pokemon can predict its total stat points! What factors are the most important in deciding this?

## Agenda:
1. Explore the data and its qualities
2. Implement train / test split with SKLearn
3. Feature engineering the categories
4. Create a linear regression model with OLS
5. Summarise the results
6. Bonus: Variance Inflation Factor (VIF)
7. Conclusion

In [1]:
# Import packages
import numpy as np                # handling maths
import pandas as pd               # handling data
import seaborn as sns             # prety data viz
import matplotlib.pyplot as plt   # basic data viz

# Import packages for train test split
from sklearn.model_selection import train_test_split

# Import stats packages for linear regression
import statsmodels.api as sm
import statsmodels.tools

# Import package for RMSE
from sklearn import metrics

# Load the dataset
pokemon_dataset = "https://github.com/digital-futures-academy/DataScienceMasterResources/raw/main/Resources/pokemon.csv"
df = pd.read_csv(pokemon_dataset)

## 1. Exploratory Data Analysis

In [2]:
# See what it looks like
pd.set_option('display.max_columns', None)
df.head(1015)

Unnamed: 0  pokedex_number               name german_name  \
0              0               1          Bulbasaur     Bisasam   
1              1               2            Ivysaur   Bisaknosp   
2              2               3           Venusaur    Bisaflor   
3              3               3      Mega Venusaur    Bisaflor   
4              4               4         Charmander    Glumanda   
...          ...             ...                ...         ...   
1010        1010             873           Frosmoth   Mottineva   
1011        1011             874        Stonjourner  Humanolith   
1012        1012             875    Eiscue Ice Face      Kubuin   
1013        1013             875  Eiscue Noice Face      Kubuin   
1014        1014             876      Indeedee Male      Servol   

            japanese_name  generation  status             species  \
0     フシギダネ (Fushigidane)           1  Normal        Seed Pokémon   
1      フシギソウ (Fushigisou)           1  Normal        Seed Pokémon   
2     フシギバナ (Fushigibana)           1  Normal        Seed Pokémon   
3     フシギバナ (Fushigibana)           1  Normal        Seed Pokémon   
4         ヒトカゲ (Hitokage)           1  Normal      Lizard Pokémon   
...                   ...         ...     ...                 ...   
1010       モスノウ (Mothnow)           8  Normal  Frost Moth Pokémon   
1011  イシヘンジン (Ishihengin)           8  Normal    Big Rock Pokémon   
1012      コオリッポ (Korippo)           8  Normal     Penguin Pokémon   
1013      コオリッポ (Korippo)           8  Normal     Penguin Pokémon   
1014       イエッサン (Yessan)           8  Normal     Emotion Pokémon   

      type_number   type_1  type_2  height_m  weight_kg  abilities_number  \
0               2    Grass  Poison       0.7        6.9                 2   
1               2    Grass  Poison       1.0       13.0                 2   
2               2    Grass  Poison       2.0      100.0                 2   
3               2    Grass  Poison       2.4      155.5                 1   
4               1     Fire     NaN       0.6        8.5                 2   
...           ...      ...     ...       ...        ...               ...   
1010            2      Ice     Bug       1.3       42.0                 2   
1011            1     Rock     NaN       2.5      520.0                 1   
1012            1      Ice     NaN       1.4       89.0                 1   
1013            1      Ice     NaN       1.4       89.0                 1   
1014            2  Psychic  Normal       0.9       28.0                 3   

        ability_1    ability_2 ability_hidden  total_points   hp  attack  \
0        Overgrow          NaN    Chlorophyll           318   45      49   
1        Overgrow          NaN    Chlorophyll           405   60      62   
2        Overgrow          NaN    Chlorophyll           525   80      82   
3       Thick Fat          NaN            NaN           625   80     100   
4           Blaze          NaN    Solar Power           309   39      52   
...           ...          ...            ...           ...  ...     ...   
1010  Shield Dust          NaN     Ice Scales           475   70      65   
1011   Power Spot          NaN            NaN           470  100     125   
1012     Ice Face          NaN            NaN           470   75      80   
1013     Ice Face          NaN            NaN           470   75      80   
1014  Inner Focus  Synchronize  Psychic Surge           475   60      65   

      defense  sp_attack  sp_defense  speed  catch_rate  base_friendship  \
0          49         65          65     45        45.0             70.0   
1          63         80          80     60        45.0             70.0   
2          83        100         100     80        45.0             70.0   
3         123        122         120     80        45.0             70.0   
4          43         60          50     65        45.0             70.0   
...       ...        ...         ...    ...         ...              ...   
10

In [3]:
df.columns.values

array(['Unnamed: 0', 'pokedex_number', 'name', 'german_name',
       'japanese_name', 'generation', 'status', 'species', 'type_number',
       'type_1', 'type_2', 'height_m', 'weight_kg', 'abilities_number',
       'ability_1', 'ability_2', 'ability_hidden', 'total_points', 'hp',
       'attack', 'defense', 'sp_attack', 'sp_defense', 'speed',
       'catch_rate', 'base_friendship', 'base_experience', 'growth_rate',
       'egg_type_number', 'egg_type_1', 'egg_type_2', 'percentage_male',
       'egg_cycles', 'against_normal', 'against_fire', 'against_water',
       'against_electric', 'against_grass', 'against_ice',
       'against_fight', 'against_poison', 'against_ground',
       'against_flying', 'against_psychic', 'against_bug', 'against_rock',
       'against_ghost', 'against_dragon', 'against_dark', 'against_steel',
       'against_fairy'], dtype=object)

In [4]:
# Seeking and dropping nulls

In [5]:
df.isnull().sum()

Unnamed: 0            0
pokedex_number        0
name                  0
german_name           0
japanese_name         0
generation            0
status                0
species               0
type_number           0
type_1                0
type_2              492
height_m              0
weight_kg             1
abilities_number      0
ability_1             3
ability_2           529
ability_hidden      232
total_points          0
hp                    0
attack                0
defense               0
sp_attack             0
sp_defense            0
speed                 0
catch_rate           18
base_friendship     115
base_experience     120
growth_rate           1
egg_type_number       0
egg_type_1            3
egg_type_2          760
percentage_male     173
egg_cycles            1
against_normal        0
against_fire          0
against_water         0
against_electric      0
against_grass         0
against_ice           0
against_fight         0
against_poison        0
against_ground  

We remove these Pokemon, as they are the only ones with nulls in these categories.

In [6]:
df.loc[df['weight_kg'].isnull(),['name']]

name
1033  Eternatus Eternamax

In [7]:
df.loc[df['egg_cycles'].isnull(),['name']]

name
658  Galarian Darmanitan Zen Mode

In [8]:
dfw = df.dropna(subset=['weight_kg','egg_cycles'], axis=0, how='any')

In [9]:
dfw.shape

(1043, 51)

What do we want to do next?
- Predict total_points against other relevant features.
- Train test split the data
- Feature engineering, picking features X and target y
- Train/fit the model
- Create a linear regression model with OLS

## 2. Train-test split

Why did we pick these features?
* Generally, we wanted to use numerical features or ones that could be easily mapped to numerical values (e.g. by one hot encoding)
* Many of the numerical features had nulls for the latest generation, such as base_friendship or base_experience
* By definition of total stat points being comprised the other stats 'attack', 'defense', 'sp_attack', 'sp_defense', 'speed', we didn't include these due to issues with multicollinearity.
* As discussed in detail in the Conclusion section at the end, there were some problems with implementing type_1 and type_2 (including multicollinearity issues), so I left these out

In [10]:
# Create the feature columns X and target y
feature_cols=['generation','status','height_m','weight_kg','hp','growth_rate','egg_cycles']
X = dfw[feature_cols]
y = dfw['total_points']

In [11]:
X.head()

generation  status  height_m  weight_kg  hp  growth_rate  egg_cycles
0           1  Normal       0.7        6.9  45  Medium Slow        20.0
1           1  Normal       1.0       13.0  60  Medium Slow        20.0
2           1  Normal       2.0      100.0  80  Medium Slow        20.0
3           1  Normal       2.4      155.5  80  Medium Slow        20.0
4           1  Normal       0.6        8.5  39  Medium Slow        20.0

In [12]:
# Run the train test split
X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    test_size = 0.2,
                                                    random_state = 42)

In [13]:
# Sanity check:
# 1. The size of the data should be consistent: y_train <> X_train, y_test <> X_test
# 2. The indices must match as well.
print(len(X_train) == len(y_train))        #should get True
print(all(X_train.index == y_train.index)) #should get true

True
True


## 3. Feature engineering

In [14]:
def feature_eng(dfw):
        dfw_local = dfw.copy()   # Good practice to make a copy!

        #One-hot encoding
        dfw_local = pd.get_dummies(dfw_local, columns = ['generation'], drop_first = True, prefix = 'generation', dtype=int)
        dfw_local = pd.get_dummies(dfw_local, columns = ['status'], drop_first = True, prefix = 'status', dtype=int)
        dfw_local = pd.get_dummies(dfw_local, columns = ['growth_rate'], drop_first = True, prefix = 'status', dtype=int)
        dfw_local = sm.add_constant(dfw_local) # Don't forget this in the context of statsmodels!
        return dfw_local

In [15]:
X_train.head()

generation         status  height_m  weight_kg  hp  growth_rate  \
544           4         Normal       0.4        7.0  49      Erratic   
256           2         Normal      10.5      740.0  75  Medium Fast   
350           3         Normal       0.8       12.0  61      Erratic   
60            1         Normal       1.0       29.5  60  Medium Fast   
929           7  Sub Legendary       3.8      100.0  83         Slow   

     egg_cycles  
544        20.0  
256        25.0  
350        15.0  
60         20.0  
929       120.0

In [16]:
## Transform the features
X_train_fe = feature_eng(X_train)
X_test_fe = feature_eng(X_test)

In [17]:
X_train_fe.head()

const  height_m  weight_kg  hp  egg_cycles  generation_2  generation_3  \
544    1.0       0.4        7.0  49        20.0             0             0   
256    1.0      10.5      740.0  75        25.0             1             0   
350    1.0       0.8       12.0  61        15.0             0             1   
60     1.0       1.0       29.5  60        20.0             0             0   
929    1.0       3.8      100.0  83       120.0             0             0   

     generation_4  generation_5  generation_6  generation_7  generation_8  \
544             1             0             0             0             0   
256             0             0             0             0             0   
350             0             0             0             0             0   
60              0             0             0             0             0   
929             0             0             0             1             0   

     status_Mythical  status_Normal  status_Sub Legendary  status_Fast  \
544                0              1                     0            0   
256                0              1                     0            0   
350                0              1                     0            0   
60                 0              1                     0            0   
929                0              0                     1            0   

     status_Fluctuating  status_Medium Fast  status_Medium Slow  status_Slow  
544                   0                   0                   0            0  
256                   0                   1                   0            0  
350                   0                   0                   0            0  
60                    0                   1                   0            0  
929                   0                   0                   0            1

In [18]:
# What are we expecting from X_train_fe?
# 1. All columns must be numerical!
print(X_train_fe.dtypes)
print('-----')

# 2. Shouldn't have any null values at this stage!
print(X_train_fe.isnull().sum())
print('-----')

# 3. The indices must still match!
print(all(X_train_fe.index == y_train.index))

const                   float64
height_m                float64
weight_kg               float64
hp                        int64
egg_cycles              float64
generation_2              int64
generation_3              int64
generation_4              int64
generation_5              int64
generation_6              int64
generation_7              int64
generation_8              int64
status_Mythical           int64
status_Normal             int64
status_Sub Legendary      int64
status_Fast               int64
status_Fluctuating        int64
status_Medium Fast        int64
status_Medium Slow        int64
status_Slow               int64
dtype: object
-----
const                   0
height_m                0
weight_kg               0
hp                      0
egg_cycles              0
generation_2            0
generation_3            0
generation_4            0
generation_5            0
generation_6            0
generation_7            0
generation_8            0
status_Mythical         0
st

## 4. OLS model

In [19]:
# Create the feature columns for the X train set
feature_cols = X_train_fe.columns

# Create a linear regression model using OLS, train the data and store the results
lin_reg = sm.OLS(y_train, X_train_fe[feature_cols])

## 5. Summarising the results

In [20]:
# Summarise the data
results = lin_reg.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           total_points   R-squared:                       0.567
Model:                            OLS   Adj. R-squared:                  0.557
Method:                 Least Squares   F-statistic:                     56.07
Date:                Tue, 04 Nov 2025   Prob (F-statistic):          1.49e-133
Time:                        23:19:16   Log-Likelihood:                -4828.2
No. Observations:                 834   AIC:                             9696.
Df Residuals:                     814   BIC:                             9791.
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
const                  372.9461     39.741      9.385      0.000     294.940     450.952
height_m                17.4435      2.743      6.360      0.000      12.060      22.827
weight_kg             2.182e-05      0.031      0.001      0.999      -0.062       0.062
hp                       1.9720      0.125     15.785      0.000       1.727       2.217
egg_cycles              -0.1169      0.267     -0.437      0.662      -0.642       0.408
generation_2           -16.6461     10.876     -1.531      0.126     -37.994       4.702
generation_3             8.1592     10.153      0.804      0.422     -11.770      28.088
generation_4             8.0032     10.485      0.763      0.446     -12.578      28.584
generation_5            -6.6810      9.470     -0.705      0.481     -25.270      11.908
generation_6             7.9232     11.812      0.671      0.503     -15.262      31.108
generation_7            -9.6931     11.616     -0.834      0.404     -32.495      13.109
generation_8            -9.3756     11.106     -0.844      0.399     -31.176      12.424
status_Mythical         21.3130     23.081      0.923      0.356     -23.993      66.619
status_Normal         -103.3295     29.479     -3.505      0.000    -161.194     -45.465
status_Sub Legendary   -15.4493     19.505     -0.792      0.429     -53.736      22.837
status_Fast            -33.7876     20.724     -1.630      0.103     -74.467       6.892
status_Fluctuating     -95.6733     31.054     -3.081      0.002    -156.629     -34.718
status_Medium Fast      -4.4981     18.497     -0.243      0.808     -40.806      31.809
status_Medium Slow       5.6283     18.845      0.299      0.765     -31.363      42.620
status_Slow             22.3397     19.328      1.156      0.248     -15.599      60.279
==============================================================================
Omnibus:                       21.573   Durbin-Watson:                   1.892
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               24.933
Skew:                          -0.325   Prob(JB):                     3.85e-06
Kurtosis:                       3.543   Cond. No.                     3.07e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.07e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

What stands out?
> R<sup>2</sup> = 0.567 - means 56.7% of the variance in the total_points is explained by the model. A decent level of predictability, but could be better.\
> F-statistic = 56.07 - model/features in X are statistically significant.\
> High condition number - means the model is still volatile, predictions may vary more when input data is changed.

In [21]:
y_pred = results.predict(X_test_fe)

In [22]:
rmse = metrics.root_mean_squared_error(y_test, y_pred)
print("Test RMSE:", rmse)

Test RMSE: 74.28572871577255


In general, this means our predictions for the total stats are about 74 stat points away from the actual values. It may be useful to calculate the normalised RMSE to get a sense of how this compares to the mean:

In [43]:
mean_total_points = df['total_points'].mean()
nrmse = rmse / mean_total_points
print("nRMSE:", round(nrmse*100,1),"%")

nRMSE: 22.6 %


So we're about 22.6% off with our predictions. Perhaps we can improve this a bit further? 

## 6. Bonus: Using Variance Inflation Factor

We can use `variance_inflation_factor` to systematically work out the best combination of features - to give us great results but without multicollineraity being an issue.

In [23]:
from statsmodels.stats.outliers_influence import variance_inflation_factor # a module to evaluate the (VIF)

cols = feature_cols

## We can create an indexed list (a series) where we list the VIF of each of the columns. Note the use of '.shape' in the second part of the loop
pd.Series([variance_inflation_factor(X_train_fe[cols].values, i) for i in range(X_train_fe[cols].shape[1])], index = X_train_fe[cols].columns)

const                   205.645788
height_m                  2.016324
weight_kg                 2.239834
hp                        1.471153
egg_cycles                9.026080
generation_2              1.439089
generation_3              1.766080
generation_4              1.497567
generation_5              1.612596
generation_6              1.342169
generation_7              1.455292
generation_8              1.439322
status_Mythical           1.938846
status_Normal            12.554423
status_Sub Legendary      2.528810
status_Fast               3.503017
status_Fluctuating        1.487605
status_Medium Fast       10.858635
status_Medium Slow        8.013404
status_Slow               9.135375
dtype: float64

In [24]:
## This a piece of code from stats.stackexchange.com

## It runs the model with all the variables.
## If any of them have a higher VIF than 10, it drops the max. 
## Then it keeps going until none of them have a higher VIF than 10.
## This leaves us with a nice set of features with no collineraity

def calculate_vif(X, thresh = 10):
    variables = list(range(X.shape[1]))
    dropped = True
    while dropped:
        dropped = False
        # this bit uses list comprehension to gather all the VIF values of the different variables
        vif = [variance_inflation_factor(X.iloc[:, variables].values, ix)
               for ix in range(X.iloc[:, variables].shape[1])]
        
        maxloc = vif.index(max(vif)) # getting the index of the highest VIF value
        if max(vif) > thresh:
            print('dropping \'' + X.iloc[:, variables].columns[maxloc] +
                  '\' at index: ' + str(maxloc))
            del variables[maxloc] # we delete the highest VIF value on condition that it's higher than the threshold
            dropped = True # if we deleted anything, we set the 'dropped' value to True to stay in the while loop

    print('Remaining variables:')
    print(X.columns[variables]) # finally, we print the variables that are still in our set
    return X.iloc[:, variables] # and return our X cut down to the remaining variables

In [25]:
X_vif = calculate_vif(X_train_fe[feature_cols])
X_vif

dropping 'const' at index: 0
dropping 'status_Normal' at index: 12
dropping 'hp' at index: 2
Remaining variables:
Index(['height_m', 'weight_kg', 'egg_cycles', 'generation_2', 'generation_3',
       'generation_4', 'generation_5', 'generation_6', 'generation_7',
       'generation_8', 'status_Mythical', 'status_Sub Legendary',
       'status_Fast', 'status_Fluctuating', 'status_Medium Fast',
       'status_Medium Slow', 'status_Slow'],
      dtype='object')


height_m  weight_kg  egg_cycles  generation_2  generation_3  \
544       0.4        7.0        20.0             0             0   
256      10.5      740.0        25.0             1             0   
350       0.8       12.0        15.0             0             1   
60        1.0       29.5        20.0             0             0   
929       3.8      100.0       120.0             0             0   
..        ...        ...         ...           ...           ...   
87        1.6      130.0        20.0             0             0   
330       1.5       55.0        15.0             0             1   
466       1.7       60.8       120.0             0             1   
121       1.5      132.5        20.0             0             0   
861       0.4        4.4        15.0             0             0   

     generation_4  generation_5  generation_6  generation_7  generation_8  \
544             1             0             0             0             0   
256             0             0             0             0             0   
350             0             0             0             0             0   
60              0             0             0             0             0   
929             0             0             0             1             0   
..            ...           ...           ...           ...           ...   
87              0             0             0             0             0   
330             0             0             0             0             0   
466             0             0             0             0             0   
121             0             0             0             0             0   
861             0             0             0             1             0   

     status_Mythical  status_Sub Legendary  status_Fast  status_Fluctuating  \
544                0                     0            0                   0   
256                0                     0            0                   0   
350                0                     0            0                   0   
60                 0                     0            0                   0   
929                0                     1            0                   0   
..               ...                   ...          ...                 ...   
87                 0                     0            0                   0   
330                0                     0            0                   0   
466                1                     0            0                   0   
121                0                     0            0                   0   
861                0                     0            0                   0   

     status_Medium Fast  status_Medium Slow  status_Slow  
544                   0                   0            0  
256                   1                   0            0  
350                   0                   0            0  
60                    1                   0            0  
929                   0                   0            1  
..                  ...                 ...          ...  
87                    0                   1            0  
330                   0                   1            0  
466                   0                   0            1  
121                   0                   0            1  
861                   1                   0            0  

[834 rows x 17 columns]

In [26]:
# Create the feature columns for the vif X train set
new_feature_cols = X_vif.columns

# Create a linear regression model using OLS, train the data and store the results
lin_reg = sm.OLS(y_train, X_train_fe[new_feature_cols])
results = lin_reg.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:           total_points   R-squared (uncentered):                   0.949
Model:                            OLS   Adj. R-squared (uncentered):              0.947
Method:                 Least Squares   F-statistic:                              885.7
Date:                Tue, 04 Nov 2025   Prob (F-statistic):                        0.00
Time:                        23:19:16   Log-Likelihood:                         -5052.8
No. Observations:                 834   AIC:                                  1.014e+04
Df Residuals:                     817   BIC:                                  1.022e+04
Df Model:                          17                                                  
Covariance Type:            nonrobust                                                  
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
height_m                32.5134      3.484      9.332      0.000      25.675      39.352
weight_kg                0.0350      0.040      0.881      0.378      -0.043       0.113
egg_cycles               1.2909      0.201      6.416      0.000       0.896       1.686
generation_2            31.1601     13.891      2.243      0.025       3.893      58.427
generation_3            79.7836     11.889      6.711      0.000      56.447     103.120
generation_4            70.7952     13.098      5.405      0.000      45.085      96.505
generation_5            40.1196     12.019      3.338      0.001      16.528      63.711
generation_6            54.2271     15.159      3.577      0.000      24.473      83.982
generation_7            28.4204     14.851      1.914      0.056      -0.730      57.571
generation_8            37.2603     14.137      2.636      0.009       9.512      65.009
status_Mythical         52.8197     26.104      2.023      0.043       1.581     104.059
status_Sub Legendary     9.8205     21.092      0.466      0.642     -31.581      51.222
status_Fast            290.2949     16.605     17.483      0.000     257.702     322.888
status_Fluctuating     232.1654     35.505      6.539      0.000     162.474     301.857
status_Medium Fast     308.1443      9.478     32.512      0.000     289.540     326.748
status_Medium Slow     318.6306     10.925     29.166      0.000     297.187     340.075
status_Slow            324.1813     14.150     22.911      0.000     296.407     351.956
==============================================================================
Omnibus:                       29.692   Durbin-Watson:                   1.975
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               75.497
Skew:                           0.048   Prob(JB):                     4.04e-17
Kurtosis:                       4.471   Cond. No.                     1.56e+03
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[3] The condition number is large, 1.56e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [27]:
X_test_vif = X_test_fe.reindex(columns=new_feature_cols, fill_value=0) # to make sure the new X_test has the same VIF columns as X_train
y_pred = results.predict(X_test_vif)

In [28]:
rmse = metrics.root_mean_squared_error(y_test, y_pred)
print("Test RMSE:", rmse)

Test RMSE: 99.24106739417145


After applying VIF, the R<sup>2</sup> value has improved drastically, meaning that a lot more of the variance is explained by the model.\
The condition number is roughly halved, but is still large enough for the model to be considered volatile.\
However, the RMSE has become significantly worse! This may suggest the model is overfitting, rather than making an improvement in the prediction as a whole.

## 7. Conclusions
Upon reflection, there are quite a few factors that I would take into account to improve this model next time:
> I could seek out a dataset that has the extra numerical values for base_experience, base_friendship, etc. for Generation 8 Pokemon, as this further information could have led to a higher RMSE score.

> I was initially concerned about incorporating the types - such as multicollinearity with many pairs like Rock/Ground, Normal/Flying, etc. Also, I would need to include both type 1 and 2 if I included any types, since from my Pokemon knowledge I am aware that many types could be underrepresented by just including type 1. For instance, the vast majority of the Flying-types would not have their Flying type represented if I did not include type 2, since most have Flying type 2. However, this would have led to incorporating a vast number of nulls in type 2 - not ideal.\
> Perhaps I could have factored this in with some sort of concatenation/duplication method that takes both types into account while ignoring nulls. One method that comes to mind is duplicating rows for each type, e.g. Charizard with Fire/Flying type gets two rows for each Fire and Flying type. This may have an issue though of overrepresenting dual-type Pokemon, but this is just one example.

# Thank you for reading this notebook! Hope you catch 'em all!

<img src="https://images.unsplash.com/photo-1665762389848-8a6acfe934c0?ixlib=rb-4.1.0&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D&auto=format&fit=crop&q=80&w=1666" width="400" height="300">

Photo by [Branden Skeli](https://unsplash.com/@branden_skeli) on [Unsplash](https://unsplash.com)